# `complete-sentences-line-comments`

Requires complete sentences for multi-line leading line comments.

## Rule Details

```ts
// Bad

// Good
```

## Options and Defaults

```json
{
  "rules": {
    "isaacscript/complete-sentences-line-comments": "error"
  }
}
```

This rule is not configurable.

## Resources

- [Rule source](../../src/rules/complete-sentences-line-comments.ts)
- [Test source](../../tests/rules/complete-sentences-line-comments.test.ts)
